pkl-codegen-java/src/main/kotlin/org/pkl/codegen/java/JavaCodeGenerator.kt

```diff
@@ -40,894 +40,931 @@ import org.pkl.core.util.CodeGeneratorUtils
+                if (firstProperty) {
+                    firstProperty = false
+                    codeBuilder.add("\$N", name)
+                } else {
+                    codeBuilder.add(", \$N", name)
+                }
+            }
+            codeBuilder.add(");\n")
+
+            methodBuilder.addCode(codeBuilder.build())
+            return methodBuilder.build()
+        }
 
-    return generateClass()
-  }
-
-  private fun generateSerialVersionUIDField(): FieldSpec {
-    return FieldSpec.builder(Long::class.java, "serialVersionUID", Modifier.PRIVATE)
-      .addModifiers(Modifier.STATIC, Modifier.FINAL)
-      .initializer("0L")
-      .build()
-  }
-
-  private fun generateEnumTypeSpec(
-    typeAlias: TypeAlias,
-    stringLiterals: Set<String>
-  ): TypeSpec.Builder {
-    val enumConstantToPklNames =
-      stringLiterals
-        .groupingBy { literal ->
-          CodeGeneratorUtils.toEnumConstantName(literal)
-            ?: throw JavaCodeGeneratorException(
-              "Cannot generate Java enum class for Pkl type alias `${typeAlias.displayName}` " +
-                "because string literal type \"$literal\" cannot be converted to a valid enum constant name."
+        fun generateSpringBootAnnotations(builder: TypeSpec.Builder) {
+            builder.addAnnotation(
+                ClassName.get("org.springframework.boot.context.properties", "ConstructorBinding")
             )
+
+            if (isModuleClass) {
+                builder.addAnnotation(
+                    ClassName.get(
+                        "org.springframework.boot.context.properties",
+                        "ConfigurationProperties"
+                    )
+                )
+            } else {
+                // not very efficient to repeat computing module property base types for every class
+                val modulePropertiesWithMatchingType =
+                    schema.moduleClass.allProperties.values.filter { property ->
+                        var propertyType = property.type
+                        while (
+                            propertyType is PType.Constrained || propertyType is PType.Nullable
+                        ) {
+                            if (propertyType is PType.Constrained) {
+                                propertyType = propertyType.baseType
+                            } else if (propertyType is PType.Nullable) {
+                                propertyType = propertyType.baseType
+                            }
+                        }
+                        propertyType is PType.Class && propertyType.pClass == pClass
+                    }
+                if (modulePropertiesWithMatchingType.size == 1) {
+                    // exactly one module property has this type -> make it available for direct
+                    // injection
+                    // (potential improvement: make type available for direct injection if it occurs
+                    // exactly
```

Comment:
That's... disappointing. I would have hoped spotless could line-wrap better [throughout]